libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteControllerFacade.kt

```diff
@@ -0,0 +1,138 @@
+            field = value
+            observers.forEach { it.onRerouteStateChanged(field) }
+        }
+
+    private val observers = CopyOnWriteArraySet<RerouteController.RerouteStateObserver>()
+
+    internal companion object {
+        private const val LOG_CATEGORY = "MapboxRerouteControllerFacade"
+
+        internal operator fun invoke(
+            rerouteObserver: RerouteControllersManager.Observer,
+            nativeRerouteController: NativeExtendedRerouteControllerInterface,
+        ): MapboxRerouteControllerFacade = MapboxRerouteControllerFacade(
+            rerouteObserver,
+            nativeRerouteController,
+        )
+    }
+
+    init {
+        nativeRerouteController.addRerouteObserver(object : RerouteObserver {
+            override fun onRerouteDetected(routeRequest: String): Boolean {
+                state = RerouteState.FetchingRoute
+                return true
+            }
+
+            override fun onRerouteReceived(
+                routeResponse: String,
+                routeRequest: String,
+                origin: RouterOrigin
+            ) {
+                state = RerouteState.RouteFetched(origin.mapToSdkRouteOrigin())
+                rerouteObserver.onNewRoutes(
+                    DirectionsResponse.fromJson(
+                        routeResponse, RouteOptions.fromUrl(URL(routeRequest))
+                    )
+                        .routes().toNavigationRoutes(origin.mapToSdkRouteOrigin()),
+                    RerouteControllersManager.Observer.Reason.Reroute,
+                )
+                state = RerouteState.Idle
+            }
+
+            override fun onRerouteCancelled() {
+                state = RerouteState.Interrupted
+                state = RerouteState.Idle
+            }
+
+            override fun onRerouteFailed(error: RerouteError) {
+                state = RerouteState.Failed(error.message)
+                state = RerouteState.Idle
+            }
+
+            override fun onSwitchToAlternative(route: RouteInterface) {
+                rerouteObserver.onNewRoutes(
+                    NavigationRoute.create(
+                        route.responseJson,
+                        route.requestUri,
+                        route.routerOrigin.mapToSdkRouteOrigin(),
+                    ),
+                    RerouteControllersManager.Observer.Reason.Alternative,
+                )
+            }
+        })
+    }
+
+    override fun reroute(callback: NavigationRerouteController.RoutesCallback) {
+        if (state is RerouteState.FetchingRoute) {
+            interrupt()
+        }
+        nativeRerouteController.setRerouteCallbackListener { expected ->
+            // listen for the very first reroute callback, most probably invoked
+            // by `nativeRerouteController.forceReroute()`. Event if no it's not a problem as a
+            // reroute request happens in any case
+            nativeRerouteController.setRerouteCallbackListener(null)
```

Comment:
this flow was added to provide for external customers option to invoke reroute at any time (our API provides it), but in NN it's a bit different: force reroute invokes reroute callback. So this way we try to set reroute callback, force reroute, and wait for the very first callback. 
It works for me, when I was testing the flow. Any objections?

Reply:
discussed with @DenisPryt. To `forceReroute` will be added callback, it will solve the concern

Reply:
Should be available in the next release [6153](https://github.com/mapbox/mapbox-navigation-native/pull/6153)